aten/src/ATen/core/typeid.h

```diff
@@ -409,49 +395,80 @@ inline bool operator!=(const TypeMeta& lhs, const TypeMeta& rhs) noexcept {
 #ifdef _MSC_VER
 #define CAFFE_KNOWN_TYPE(T)                                               \
   template <>                                                             \
-  AT_CORE_EXPORT TypeIdentifier TypeMeta::Id<T>() {                       \
+  AT_CORE_EXPORT TypeIdentifier TypeIdentifier::Get<T>() {                \
     static const TypeIdentifier type_id = TypeIdentifier::createTypeId(); \
-    static TypeNameRegisterer<T> registerer(type_id, #T);                 \
     return type_id;                                                       \
+  }                                                                       \
+  namespace detail {                                                      \
+  template <>                                                             \
+  AT_CORE_EXPORT const char* __TypeName<T>() noexcept {                   \
+    return #T;                                                            \
+  }                                                                       \
+  template <>                                                             \
+  AT_CORE_EXPORT const TypeMetaData* _TypeMetaData<T>() noexcept {        \
+    static TypeMetaData singleton = createTypeMetaData<T>();              \
+    return &singleton;                                                    \
+  }                                                                       \
   }
 #else // _MSC_VER
 #define CAFFE_KNOWN_TYPE(T)                                               \
   template <>                                                             \
-  TypeIdentifier TypeMeta::Id<T>() {                                      \
+  TypeIdentifier TypeIdentifier::Get<T>() {                               \
     static const TypeIdentifier type_id = TypeIdentifier::createTypeId(); \
-    static TypeNameRegisterer<T> registerer(type_id, #T);                 \
     return type_id;                                                       \
+  }                                                                       \
+  namespace detail {                                                      \
+  template <>                                                             \
+  const char* __TypeName<T>() noexcept {                                  \
+    return #T;                                                            \
+  }                                                                       \
+  template <>                                                             \
+  const TypeMetaData* _TypeMetaData<T>() noexcept {                       \
```

Comment:
Ummm... this seems very dodgy.

https://stackoverflow.com/questions/15061774/template-class-member-specialization-without-declaration-in-header suggests that the C++ standard states that specializations of C++ templates must be declared before they are used. The classic usage of `CAFFE_KNOWN_TYPE` is inside a C++ file, so these specializations are never declared outside of the object files. So you are violating the standard when you expect this to work.

Reply:
(This doesn't seem like what is causing the CI failure though...)

Reply:
OK, I guess this is what the code was doing before, so it probably should work.

Reply:
Yes, it is dodgy but basically the only way we got it to work reliably across translation units together with python. As you said, this pattern was there before, so shouldn't make a difference here.